src/client/api/init.c

```diff
@@ -354,7 +356,6 @@ daos_fini(void)
 	dc_job_fini();
 
 	pl_fini();
-	daos_hhash_fini();
```

Comment:
If daos_hhash_fini() is not moved ahead, then the pool/container might be hold by dc_object by the following call sequence, i.e. the dc_pool/cont will only be finally freed in daos_hhash_fini(), though dc_pool_fini()/dc_cont_fini() is already executed which suppose to cleanup everything about dc_pool/cont, so dc_pool_free might cause issue here, 

daos_hhash_fini()->d_hash_table_destory_inplace->d_hash_rec_delete_at->ch_rec_free->hh_op_rec_free->obj_free->dc_pool_put/dc_cont_put

Reply:
@mchaarawi  please check, thanks.

Reply:
so that is a bug though right? it sounds like a leak where someone left a hashtable that was not destroyed with open handles? how. do we reproduce?

Reply:
Im going to push a revert to this PR since the question is not answered and test case was not added.

Reply:
https://github.com/daos-stack/daos/pull/15621

Reply:
> it sounds like a leak where someone left a hashtable that was not destroyed with open handles?

@mchaarawi  Sorry, missed your question.

daos_hhash_fini() is to make sure all the allocated instance to be freed before dc_obj_fini()/dc_cont_fini/dc_pool_fini() are called. NB: these dc_xxx  cleanup the module related structures in the client stack. IMHO, it makes sense to cleanup all instances before cleanup the module/layer. 

TBH this looks more like a typo to me, adding a test in this case seems not really help much. or I misunderstood your concern?

Reply:
> > it sounds like a leak where someone left a hashtable that was not destroyed with open handles?
> 
> @mchaarawi Sorry, missed your question.
> 
> daos_hhash_fini() is to make sure all the allocated instance to be freed before dc_obj_fini()/dc_cont_fini/dc_pool_fini() are called. NB: these dc_xxx cleanup the module related structures in the client stack. IMHO, it makes sense to cleanup all instances before cleanup the module/layer.
> 

i don't see how this is a typo. we are freeing things in the order they are initialized. what im saying is that if there is something that is leaked and not closed, we should fix that and not require the hash fini to close it. otherwise the hash should not close objects that are still open. so again please provide a test / example for this.

> TBH this looks more like a typo to me, adding a test in this case seems not really help much. or I misunderstood your concern?

TBH typo or not, any bug fix should be accompanied by a regression test. it will also help clarify the situation that you are trying to fix.

Reply:
If you can provide the sequence of things you did to reproduce the segfault, i can take a look too.